llama-rs/src/llama.rs

```diff
@@ -591,54 +630,59 @@ impl LlamaModel {
     }
 
     pub fn inference_with_prompt<'a>(
-        &self,
+        &mut self,
         vocab: &'a GptVocab,
         params: &InferenceParams,
         prompt: &str,
         rng: &mut impl rand::Rng,
+        stop_after_prompt: bool,
         callback: impl Fn(OutputToken<'a>),
     ) {
-        let embd_inp = self.tokenize(vocab, prompt, true);
-        let mut logits = Vec::new();
+        let beginning_of_sentence = self.n_past == 0;
+        let embd_inp = self.tokenize(vocab, prompt, beginning_of_sentence);
+        let mut logits = vec![0.0; self.hparams.n_vocab as usize];
 
-        // determine the required inference memory per token:
+        // Mem per token is initialized at zero. Will be read the first time
+        // llama_eval is called.
         let mut mem_per_token = 0;
-        let _ = self.llama_eval(
-            params.n_threads,
-            0,
-            &[0, 1, 2, 3],
-            &mut logits,
-            &mut mem_per_token,
-        );
 
-        let last_n_size = params.repeat_last_n;
-        let mut last_n_tokens = vec![0 as TokenId; last_n_size];
+        unsafe {
+            self.get_memory();
+        }
+
+        let mut last_n_tokens = vec![0 as TokenId; params.repeat_last_n];
 
         let mut remaining_tokens = usize::min(
             params.n_predict,
             self.hparams.n_ctx as usize - embd_inp.len(),
         );
         let mut input_consumed = 0;
 
-        let mut n_past = 0;
         let mut embd = Vec::new();
         while remaining_tokens > 0 {
             // predict
             if embd.len() > 0 {
                 self.llama_eval(
                     params.n_threads,
-                    n_past,
+                    self.n_past as i32,
                     &embd,
                     &mut logits,
                     &mut mem_per_token,
                 );
             }
 
-            n_past += embd.len() as i32;
+            self.n_past += embd.len();
```

Comment:
`n_past` is now stored inside the `LlamaModel` object. This allows invoking the generation multiple times. Wen n_past is > 0, inference continues by adding any extra prompt you feed it, and then starts inferring normally.

This could also serve as the basis for a more complex chat-like interaction :)